sw/device/tests/rv_core_ibex_address_translation_test.c

```diff
@@ -0,0 +1,204 @@
+      &flash_ctrl,
+      mmio_region_from_addr(TOP_EARLGREY_FLASH_CTRL_CORE_BASE_ADDR)));
+  flash_ctrl_testutils_wait_for_init(&flash_ctrl);
+
+  // Set up default access for data partitions.
+  flash_ctrl_testutils_default_region_access(
+      &flash_ctrl, /*rd_en=*/true, /*prog_en=*/true, /*erase_en=*/true,
+      /*scramble_en=*/false, /*ecc_en=*/false, /*high_endurance_en=*/false);
+
+  return flash_ctrl;
+}
+
+bool test_main(void) {
+  // Calculate 2-byte aligned addresses to put a copy of the toy functions in.
+  uintptr_t flash_mem_end_addr = TOP_EARLGREY_FLASH_CTRL_MEM_BASE_ADDR +
+                                 TOP_EARLGREY_FLASH_CTRL_MEM_SIZE_BYTES;
+  uintptr_t offset = 0x1000;
+  uintptr_t make_lower_case_addr = flash_mem_end_addr - offset;
+  uintptr_t get_name_addr = flash_mem_end_addr - (2 * offset);
+
+  // Initiate the flash controller.
+  dif_flash_ctrl_state_t flash_ctrl = init_flash();
+
+  // Move the toy functions to their aligned addresses.
+  CHECK(!flash_ctrl_testutils_erase_and_write_page(
+      /*flash_state=*/&flash_ctrl,
+      /*byte_address=*/(uint32_t)make_lower_case_addr,
+      /*partition_id=*/0,
+      /*data=*/(uint32_t *)make_lower_case,
+      /*partition_type=*/kDifFlashCtrlPartitionTypeData,
+      /*word_count=*/kMakeLowerCaseFnSize / 4));
+  CHECK(!flash_ctrl_testutils_erase_and_write_page(
+      /*flash_state=*/&flash_ctrl,
+      /*byte_address=*/(uint32_t)get_name_addr,
+      /*partition_id=*/0,
+      /*data=*/(uint32_t *)get_name,
+      /*partition_type=*/kDifFlashCtrlPartitionTypeData,
+      /*word_count=*/kGetNameFnSize / 4));
+
+  char test_str[] = TEST_STR;
+  ((str_fn_t)make_lower_case_addr)(test_str);
+  CHECK_STR_EQ(test_str, EXPECTED_RESULT_MAKE_LOWER_CASE);
+
+  ((str_fn_t)get_name_addr)(test_str);
+  CHECK_STR_EQ(test_str, EXPECTED_RESULT_GET_NAME);
+
+  // The memory address to which the functions will be mapped.
+  const uintptr_t kRemapAddr = 0xA0000000;
+
+  // Create translation descriptions.
+  dif_rv_core_ibex_addr_translation_mapping_t make_lower_case_mapping = {
+      .matching_addr = kRemapAddr,
+      .remap_addr = make_lower_case_addr,
+      .size = next_power_of_two(kMakeLowerCaseFnSize),
+  };
+  dif_rv_core_ibex_addr_translation_mapping_t get_name_mapping = {
+      .matching_addr = kRemapAddr,
+      .remap_addr = get_name_addr,
+      .size = next_power_of_two(kGetNameFnSize),
+  };
+  const dif_rv_core_ibex_addr_translation_bus_t kBuses[] = {
+      kDifRvCoreIbexAddrTranslationIBus, kDifRvCoreIbexAddrTranslationDBus};
+  const dif_rv_core_ibex_addr_translation_slot_t kSlots[] = {
+      kDifRvCoreIbexAddrTranslationSlot_0, kDifRvCoreIbexAddrTranslationSlot_1};
+
+  // Get ibex core handle.
+  dif_rv_core_ibex_t ibex_core;
+  CHECK_DIF_OK(dif_rv_core_ibex_init(
+      mmio_region_from_addr(TOP_EARLGREY_RV_CORE_IBEX_CFG_BASE_ADDR),
+      &ibex_core));
+
+  // Map virtual address space to make_lower_case() using slot 1.
```

Comment:
Seems you run almost the same code but mapping to different slots. Move it to a function and pass the slot?

Reply:
Agreed. I've wrapped up this repetitive in functions.